python/strings_udf/cpp/include/cudf/strings/udf/split.cuh

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "udf_string.cuh"
+
+#include <cudf/strings/detail/split_utils.cuh>
+#include <cudf/strings/string_view.cuh>
+
+namespace cudf {
+namespace strings {
+namespace udf {
+
+/**
+ * @brief Split string using given string
+ *
+ * The caller must allocate an array of cudf::string_view to be filled
+ * in by this function. Calling this with a `nullptr` for the `result`
+ * will return the number of elements.
+ *
+ * @code{.cpp}
+ * auto d_str = cudf::string_view{"the best  of times ", 19};
+ * auto tgt = cudf::string_view{}; // empty string
+ * auto token_count = split(d_str, tgt, nullptr);
+ * auto result = new cudf::string_view[token_count];
+ * split(d_str, tgt, result);
+ * // result is array like ["the", "best", "of", "times"]
+ * @endcode
+ *
+ * @param d_str String to split
+ * @param tgt String to split on
+ * @param result Empty array to populate with output objects.
+ *               Pass `nullptr` to just get the token count.
```

Comment:
Is this the most efficient way to get the token count, or is there another function that can be used to count the number of instances of a split token? The two pass approach of this API may not be necessary if another function can be used for the split token counting.

Reply:
There is no more efficient way. I could create a separate count-tokens function but the logic would only be different by one line of code. If you think that would be a better interface, I could refactor to make a utility function out of the current one.

Reply:
There is not an existing API that can count tokens for this. 
This is loosely based on [std::strtok](https://en.cppreference.com/w/cpp/string/byte/strtok) though is probably not the best behavior to emulate. It is also similar to how CUB APIs work too but still.
I would be more concerned if this was a public API instead of an internal utility for the python udf code.
I will give it some consideration.